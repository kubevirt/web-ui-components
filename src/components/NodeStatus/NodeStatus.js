import React from 'react';
import { Popover } from 'patternfly-react';
import PropTypes from 'prop-types';

import {
  getNodeStatus,
  NODE_STATUS_UNDER_MAINTENANCE,
  NODE_STATUS_STOPPING_MAINTENANCE,
} from '../../utils/status/node';
import { getCreationTimestamp, getDeletionTimestamp, getName, getMaintenanceReason } from '../../selectors';
import { OverlayStatus } from '../Status';

const UnderMaintenanceStatus = ({ node, maintenance, TimestampComponent }) => {
  const maintenanceReason = getMaintenanceReason(maintenance);
  const created = getCreationTimestamp(maintenance);
  const overlay = (
    <Popover id={`${getName(node)}-status-popover`} title="Under maintenance">
      <div>This host is under maintenance.</div>
      {maintenanceReason && (
        <React.Fragment>
          <br />
          <b>Maintenance reason:</b>
          <div>{maintenanceReason}</div>
        </React.Fragment>
      )}
      <br />
      <div>
        Started: <TimestampComponent simple timestamp={created} />
      </div>
    </Popover>
  );

  return <OverlayStatus icon="off" overlay={overlay} text="Under maintenance" />;
};

UnderMaintenanceStatus.defaultProps = {
  maintenance: null,
};

UnderMaintenanceStatus.propTypes = {
  node: PropTypes.object.isRequired,
  maintenance: PropTypes.object,
  TimestampComponent: PropTypes.oneOfType([PropTypes.node, PropTypes.func]).isRequired,
};

const StoppingMaintenanceStatus = ({ node, maintenance, TimestampComponent }) => {
  const created = getCreationTimestamp(maintenance);
  const deleted = getDeletionTimestamp(maintenance);
  const overlay = (
    <Popover id={`${getName(node)}-status-popover`} title="Stopping maintenance">
      <div>This host is leaving maintenance. It will rejoin the cluster and resume accepting workloads.</div>
      <br />
      <div>
        Started: <TimestampComponent simple timestamp={created} />
      </div>
      <div>
        Ended: <TimestampComponent simple timestamp={deleted} />
      </div>
    </Popover>
  );

  return <OverlayStatus icon="off" overlay={overlay} text="Stopping maintenance" />;
};

StoppingMaintenanceStatus.defaultProps = {
  maintenance: null,
};

StoppingMaintenanceStatus.propTypes = {
  node: PropTypes.object.isRequired,
  maintenance: PropTypes.object,
  TimestampComponent: PropTypes.oneOfType([PropTypes.node, PropTypes.func]).isRequired,
};

export const NodeStatus = ({ node, maintenances, TimestampComponent }) => {
  const nodeStatus = getNodeStatus(node, maintenances);
  switch (nodeStatus.status) {
    case NODE_STATUS_UNDER_MAINTENANCE:
      return (
        <UnderMaintenanceStatus
          node={node}
          maintenance={nodeStatus.maintenance}
          TimestampComponent={TimestampComponent}
        />
      );
    case NODE_STATUS_STOPPING_MAINTENANCE:
      return (
        <StoppingMaintenanceStatus
          node={node}
          maintenance={nodeStatus.maintenance}
          TimestampComponent={TimestampComponent}
        />
      );
    default:
      return false;
  }
};

NodeStatus.defaultProps = {
  maintenances: null,
};

NodeStatus.propTypes = {
  node: PropTypes.object.isRequired,
  maintenances: PropTypes.array,
  TimestampComponent: PropTypes.oneOfType([PropTypes.node, PropTypes.func]).isRequired,
};
