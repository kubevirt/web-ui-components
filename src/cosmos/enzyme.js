import createContext from 'react-cosmos-test/enzyme';

import paths from '../../config/paths';

export default function createTestContext(args) {
  args.proxies = args.proxies || []; // workaround of react-cosmos-test bug
  return createContext({
    ...args,
    cosmosConfigPath: `${paths.config}/cosmos.config.js`,
  });
}
