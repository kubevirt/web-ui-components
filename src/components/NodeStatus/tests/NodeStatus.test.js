import React from 'react';
import { shallow } from 'enzyme';

import { NodeStatus } from '../NodeStatus';
import NodeStatusFixtures from '../fixtures/NodeStatus.fixture';

const testNodeStatus = () => <NodeStatus {...NodeStatusFixtures[0].props} />;

describe('<NodeStatus />', () => {
  it('renders correctly', () => {
    const component = shallow(testNodeStatus());
    expect(component).toMatchSnapshot();
  });
});
