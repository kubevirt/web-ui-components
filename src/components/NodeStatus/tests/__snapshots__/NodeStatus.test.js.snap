// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NodeStatus /> renders correctly 1`] = `
<UnderMaintenanceStatus
  TimestampComponent={[Function]}
  maintenance={
    Object {
      "apiVersion": "kubevirt.io/v1alpha1",
      "kind": "NodeMaintenance",
      "metadata": Object {
        "creationTimestamp": "2019-04-25T18:49:13Z",
        "name": "nodemaintenance-fooNodeName",
      },
      "spec": Object {
        "nodeName": "fooNodeName",
        "reason": "some reason",
      },
    }
  }
  node={
    Object {
      "metadata": Object {
        "name": "fooNodeName",
      },
    }
  }
/>
`;
